File: packages/auth/amplify_auth_cognito_ios/ios/Classes/Utils/UserAttributeDeserialization.swift
/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import Foundation
import Amplify

func createAuthUserAttribute(key: String, value: String) -> AuthUserAttribute {
    let authUserAttributeKey = createAuthUserAttributeKey(keyName: key);
    return AuthUserAttribute(authUserAttributeKey, value: value)
}

func createAuthUserAttributeKey(keyName: String) -> AuthUserAttributeKey {
    switch keyName {
    case "address":
        return AuthUserAttributeKey.address
    case "birthdate":
        return AuthUserAttributeKey.birthDate
    case "email":
        return AuthUserAttributeKey.email
    case "family_name":
        return AuthUserAttributeKey.familyName
    case "gender":
        return AuthUserAttributeKey.gender
    case "given_name":
        return AuthUserAttributeKey.givenName
    case "locale":
        return AuthUserAttributeKey.locale
    case "middle_name":
        return AuthUserAttributeKey.middleName
    case "name":
        return AuthUserAttributeKey.name
    case "nickname":
        return AuthUserAttributeKey.nickname
    case "phone_number":
        return AuthUserAttributeKey.phoneNumber
    case "picture":
        return AuthUserAttributeKey.picture
    case "preferred_username":
        return AuthUserAttributeKey.preferredUsername
    default:
        // amplify-ios does not currently include enums for these keyNames
        let unknownKeys = ["profile", "updated_at", "website", "zoneinfo"]
        if (unknownKeys.contains(keyName)) {
            return AuthUserAttributeKey.unknown(keyName)
        } else {
            // amplify-ios prepends 'custom:' to custom keys
            let customKeyPrefix = "custom:"
            let customKeyName = keyName.starts(with: customKeyPrefix) ? String(keyName.dropFirst(customKeyPrefix.count)) : keyName
            return AuthUserAttributeKey.custom(customKeyName)
        }
    }
}

File: packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/utils/UserAttributeDeserialization.kt
/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

package com.amazonaws.amplify.amplify_auth_cognito.utils

import com.amplifyframework.auth.AuthUserAttribute
import com.amplifyframework.auth.AuthUserAttributeKey

fun createAuthUserAttribute(key: String, value: String): AuthUserAttribute {
    val authUserAttributeKey: AuthUserAttributeKey = createAuthUserAttributeKey(key)
    return AuthUserAttribute(authUserAttributeKey, value)
}

fun createAuthUserAttributeKey(keyName: String): AuthUserAttributeKey {
    when (keyName) {
        "address" -> return AuthUserAttributeKey.address()
        "birthdate" -> return AuthUserAttributeKey.birthdate()
        "email" -> return AuthUserAttributeKey.email()
        "family_name" -> return AuthUserAttributeKey.familyName()
        "gender" -> return AuthUserAttributeKey.gender()
        "given_name" -> return AuthUserAttributeKey.givenName()
        "locale" -> return AuthUserAttributeKey.locale()
        "middle_name" -> return AuthUserAttributeKey.middleName()
        "name" -> return AuthUserAttributeKey.name()
        "nickname" -> return AuthUserAttributeKey.nickname()
        "phone_number" -> return AuthUserAttributeKey.phoneNumber()
        "picture" -> return AuthUserAttributeKey.picture()
        "preferred_username" -> return AuthUserAttributeKey.preferredUsername()
        "profile" -> return AuthUserAttributeKey.profile()
        "updated_at" -> return AuthUserAttributeKey.updatedAt()
        "website" -> return AuthUserAttributeKey.website()
        "zoneinfo" -> return AuthUserAttributeKey.zoneInfo()
        else -> return createCustomAuthUserAttributeKey(keyName)
    }
}

private fun createCustomAuthUserAttributeKey(keyName: String): AuthUserAttributeKey {
    val customKey: String = if (keyName.startsWith("custom:")) keyName else "custom:$keyName"
    return AuthUserAttributeKey.custom(customKey)
}
